pkg/cloud/services/autoscaling/autoscalinggroup_test.go

```diff
@@ -220,3 +317,722 @@ func TestService_SDKToAutoScalingGroup(t *testing.T) {
+				if err != nil {
+					t.Fatalf("did not expect error: %v", err)
+				}
+				if autoscalinggroup != nil {
+					t.Fatalf("expected asg to be nil but got: %+v", autoscalinggroup)
+				}
+			},
+		},
+		{
+			name:            "should return error if create ASG fails",
+			machinePoolName: "create-asg-fail",
+			expect: func(s *mock_cloudiface.MockClusterScoperMockRecorder, m *mock_autoscalingiface.MockAutoScalingAPIMockRecorder) {
+				s.Name().Times(2).Return("test")
+				s.Info(gomock.Eq("Running instance"))
+				m.CreateAutoScalingGroup(gomock.AssignableToTypeOf(&autoscaling.CreateAutoScalingGroupInput{})).Return(nil, awserrors.NewFailedDependency("dependency failure"))
+				s.Error(gomock.Any(), gomock.Eq("unable to create AutoScalingGroup"))
+			},
+			check: func(autoscalinggroup *expinfrav1.AutoScalingGroup, err error) {
+				if err == nil {
+					t.Fatalf("expected error to occur")
+				}
+				if autoscalinggroup != nil {
+					t.Fatalf("expected asg to be nil but got: %+v", autoscalinggroup)
+				}
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			asgMock := mock_autoscalingiface.NewMockAutoScalingAPI(mockCtrl)
+			scopeMock := mock_cloudiface.NewMockClusterScoper(mockCtrl)
+			tt.expect(scopeMock.EXPECT(), asgMock.EXPECT())
+			s := &Service{
+				scope:     scopeMock,
+				ASGClient: asgMock,
+			}
+			mps, err := getMachinePoolScope(tt.machinePoolName)
+			if err != nil {
+				t.Fatalf("Failed to create context: %v", err)
+			}
+			asg, err := s.CreateASG(mps)
+			tt.check(asg, err)
+		})
+	}
+}
+
+func TestService_UpdateASG(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	tests := []struct {
+		name            string
+		machinePoolName string
+		expect          func(s *mock_cloudiface.MockClusterScoperMockRecorder, m *mock_autoscalingiface.MockAutoScalingAPIMockRecorder)
+		check           func(err error)
+	}{
+		{
+			name:            "should return without error if update ASG is successful",
+			machinePoolName: "update-asg-success",
+			expect: func(s *mock_cloudiface.MockClusterScoperMockRecorder, m *mock_autoscalingiface.MockAutoScalingAPIMockRecorder) {
+				m.UpdateAutoScalingGroup(gomock.AssignableToTypeOf(&autoscaling.UpdateAutoScalingGroupInput{})).Return(&autoscaling.UpdateAutoScalingGroupOutput{}, nil)
+			},
+			check: func(err error) {
+				if err != nil {
+					t.Fatalf("did not expect error: %v", err)
+				}
+			},
+		},
+		{
+			name:            "should return error if update ASG fails",
+			machinePoolName: "update-asg-fail",
+			expect: func(s *mock_cloudiface.MockClusterScoperMockRecorder, m *mock_autoscalingiface.MockAutoScalingAPIMockRecorder) {
+				m.UpdateAutoScalingGroup(gomock.AssignableToTypeOf(&autoscaling.UpdateAutoScalingGroupInput{})).Return(nil, awserrors.NewFailedDependency("dependency failure"))
+			},
+			check: func(err error) {
+				if err == nil {
+					t.Fatalf("expected error to occur")
+				}
+			},
+		},
```

Comment:
Can we also add scenario:
Should fetch subnets from infra cluster if no subnets specified in AWSMachinePool